react_juce/core/CanvasView.h

```diff
@@ -57,613 +57,495 @@ namespace blueprint
-                        jassert(args.numArguments == 0);
-                        juce::ignoreUnused(args);
+            const int x = args.arguments[0];
+            const int y = args.arguments[1];
+            const int width = args.arguments[2];
+            const int height = args.arguments[3];
 
-                        // Reset the current path on a call to beginPath on the context.
-                        path = juce::Path();
+            g.setColour(ctx.properties.strokeStyle.colour);
+            g.drawRect(x, y, width, height, ctx.properties.lineWidth);
+        }
 
-                        return juce::var();
-                    }
-            });
+        void strokeRoundedRect(CanvasContextState &ctx, juce::Graphics &g, const juce::var::NativeFunctionArgs &args)
+        {
+            jassert(args.numArguments == 5);
 
-            setProperty("lineTo", juce::var::NativeFunction {
-                    [=](const juce::var::NativeFunctionArgs& args) -> juce::var {
-                        jassert(graphics);
-                        jassert(args.numArguments == 2);
+            const float x = args.arguments[0];
+            const float y = args.arguments[1];
+            const float width = args.arguments[2];
+            const float height = args.arguments[3];
+            const float cornerSize = args.arguments[4];
 
-                        const float x = args.arguments[0];
-                        const float y = args.arguments[1];
+            g.setColour(ctx.properties.strokeStyle.colour);
+            g.drawRoundedRectangle(x, y, width, height, cornerSize, (float)ctx.properties.lineWidth);
+        }
 
-                        path.lineTo(x, y);
+        void fillRoundedRect(CanvasContextState &ctx, juce::Graphics &g, const juce::var::NativeFunctionArgs &args)
+        {
+            jassert(args.numArguments == 5);
 
-                        return juce::var();
-                    }
-            });
+            const float x = args.arguments[0];
+            const float y = args.arguments[1];
+            const float width = args.arguments[2];
+            const float height = args.arguments[3];
+            const float cornerSize = args.arguments[4];
 
-            setProperty("moveTo", juce::var::NativeFunction {
-                    [=](const juce::var::NativeFunctionArgs& args) -> juce::var {
-                        jassert(graphics);
-                        jassert(args.numArguments == 2);
-                        juce::ignoreUnused(args);
+            g.setFillType(ctx.properties.fillStyle);
+            g.fillRoundedRectangle(x, y, width, height, cornerSize);
+        }
 
-                        const float x = args.arguments[0];
-                        const float y = args.arguments[1];
+        void clearRect(CanvasContextState &ctx, juce::Graphics &g, const juce::var::NativeFunctionArgs &args)
+        {
+            //TODO: How are we implementing this without an image?
```

Comment:
We could still do this via image if needed, right? instead of passing the actual `juce::Graphics` instance that we get in `paint` we could make a `juce::Image`, construct a graphics instance pointing to that image, run all the context/draw operations, then paint the image to the real graphics context?

I don't know how necessary that is, but seems like an option

Reply:
Yeah we could do that indeed. And from memory I think there may be actually functional reasons for doing this related to the old `autoClear` prop. Removing it does nicely solve the scalling issue on retina :-). 

I'm quite happy to render to an Image still and then fix the scalling stuff separately. There was a state based reason for rendering to an `Image` from what I remember? Something about scrolling waveforms etc.

Reply:
Ah yea, for a scrolling waveform you could, for example, snapshot the current state, paint it back to the canvas but shifted 1px to the left, and then explicitly paint only the rightmost 1px region of the canvas.

So... yea... might want the image there :D. The retina scaling issue is annoying though... I wonder if we could solve it by making that `juce::Image` 2x as big as the actual component and render all these draw commands onto that but automatically scale the positions by 2? Then when painting the image to the screen we just scale it then? That would address the retina @2x part right?

Reply:
Yeah totally. I did attempt something along those lines a while back and got most of it working I think. OK cool. I'll continue rendering to an Image and reinstate that `autoclear` behaviour. I think we should setup a default prop for that set to `true` maybe?

Reply:
:+1: sounds good. I'm tempted to default `autoclear` to `false`, it feels like a "favor explicit instead of implicit" type of argument. That is, if we autoclear by default, then we have this implicit behavior that occurs in between two subsequent calls to your onDraw callback. If we don't autoclear by default, then nothing happens between two subsequent calls, there's no surprise there, and if you want clearing behavior you can explicitly request it with a 1 prop change

Reply:
OK cool sounds reasonable to me. Will fix this up today and finish things off.

Reply:
Actually thinking about this it's really useful to have a C++ `CanvasContext` equivalent exposed in the `CanvasView` header. 

With the old implementation I used this to allow me to push look and feel paint overrides into JS on some C++ components. i.e. a hybrid between a C++ and React component. This is useful for very specific components which blueprint doesn't provide equivalent functionality for right now but which we still want to "skin" from React. 

How would you feel about keeping a public function available similar to the following:

`DrawCanvasCommands(CanvasContext &ctx, const juce::var &drawCommands)`

In the above `CanvasContext` will probably just be a struct or similar to the original implementation (i.e. holds image and graphics context) and `drawCommands` is the JS command array passed into native.

Reply:
Thinking about it. Might make more sense from API point of view to have a 

`CanvasContext::processDrawCommands(const juce::var &drawCommands)` ?

Reply:
Hmm interesting, so you're thinking something like the following?

```cpp
class CanvasView : public blueprint::View
{
        struct CanvasContext
        {
            int                                width  = 1;
            int                                height = 1;
            juce::Path                         path;
            CanvasContextProperties            properties;
            std::vector<juce::AffineTransform> transformStack;
            juce::Image internalImage;
            
            void processDrawCommands(const juce::var &drawCommands);
        };

        // Regular canvasView stuff here etc etc

private:
       CanvasContext ctx;
};
```

And with this, CanvasView's implementation largely composes over the behavior of the internal `CanvasContext` struct, which is made public to accommodate situations like the one you have?

I can get into this idea, seems like a clean way API for the CanvasView itself anyways

Reply:
Yep exactly. Something along these lines. I think we'd still have an anonymous std::map of command -> function but it would look more like:

```
void setFillStyle(CanvasContextState &ctx, const juce::var::NativeFunctionArgs &args)
{
    //DO necessary work with ctx. Possibly just ctx.setFillStyle()?
}

std::unordered_map<juce::String, DrawCommand> DrawCommands
{
    {"setFillStyle", setFillStyle}
}

void CanvasContext::processDrawCommands(const juce::var &drawCommands)
{
    jassert(drawCommands.isArray());                                                                                                                                                                 
                                                                                                                                                                                                                   
    for (juce::var &command : *drawCommands.getArray())                                                                                                                                              
    {                                                                                                                                                                                                
        jassert(command.isArray());                                                                                                                                                                  
        jassert(command.getArray()->size() >= 1);                                                                                                                                                    
                                                                                                                                                                                                                   
        const auto args = command.getArray();                                                                                                                                                        
        const auto name = (*args)[0];                                                                                                                                                                
                                                                                                                                                                                                                   
        if (auto it = DrawCommands.find(name); it != DrawCommands.end())                                                                                                                             
        {                          
            // CanvasContext passes self to the existing static functions in our DrawCommands map
            // Could take a different approach but this is probably lighter than each CanvasContext object
            // holding a map of strings to functions                                                                                                                                                                  
            it->second(this, g, juce::var::NativeFunctionArgs(juce::var(), args->getRawDataPointer() + 1, args->size() - 1));                                                                         
        }                                                                                                                                                                                            
    } 
}
```

Reply:
Yup, sounds good. Last thought:

```
void setFillStyle(CanvasContextState &ctx, const juce::var::NativeFunctionArgs &args)
{
    //DO necessary work with ctx. Possibly just ctx.setFillStyle()?
}
```

I think I'd prefer to have these static functions just be kind of pure functions that operate over the CanvasContext and make the necessary mutations. It seems like it'd end up being a lot of duplication to write a static `setFillStyle` that just unwraps the args and then calls _back_ to `ctx.setFillStyle`

Reply:
Yeah cool. That was a choice to make. In which case these functions will probably need to just directly access the `CanvasContext`'s `juce::Graphics` instance? That graphics instance will be setup to render to the `juce::Image` in `CanvasContext`'s constructor I guess. 

Cool with that general approach?

Reply:
Ah yea, yup, that sounds ok to me!

Reply:
Sweeeeet. Let's finish this bad boy.